### PR TITLE
Let Slack users inherit certain Kifi bot scopes

### PR DESCRIPTION
**Pull request by leogrim**
@ryanpbrewster

### CODE REVIEW

**Comment by ryanpbrewster** [comment]
Reading this on my phone is hard, but most of it looks nice.

**Comment by leogrim** [comment]
What do you think next of being able to give multiple tokens to some of the `SlackClientWrapper` methods, with a `collectFirst` that works approach?

File: kifi-backend/modules/common/app/com/keepit/slack/models/SlackAuth.scala [hunk 1]
```diff
@@ -48,6 +48,8 @@ object SlackAuthScope {
   val UsersRead = SlackAuthScope("users:read")
   val UsersWrite = SlackAuthScope("users:write")
 
+  val inheritableBotScopes: Set[SlackAuthScope] = Set(Bot, UsersRead, TeamRead, ChannelsRead) // scopes covering APIs that can use the Kifi bot token transparently
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/slack/models/SlackAuth.scala:51]: `userEquivalentBotScopes`?

File: kifi-backend/modules/shoebox/app/com/keepit/slack/SlackCommander.scala [hunk 2]
```diff
@@ -123,8 +110,48 @@ class SlackCommanderImpl @Inject() (
...
+      kifiBotToken <- slackTeam.kifiBotToken
+    } yield kifiBotToken
+
+    kifiBotTokenOpt match {
+      case Some(kifiBotToken) => slackClient.validateToken(kifiBotToken).imap {
+        case true => SlackAuthScope.inheritableBotScopes
+        case false => Set.empty[SlackAuthScope]
+      }
+      case None => Future.successful(Set.empty[SlackAuthScope])
+    }
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/SlackCommander.scala:137]: I think a `map().getOrElse` or a `fold` is cleaner here

**Comment by ryanpbrewster** [comment]
I'm fine doing a `collectFirst` on a sequence of tokens starting with the `kifiBotToken` if it exists, but that should be done within each method IMO. I like APIs with convention
`searchMessagesViaUser(membership: SlackTeamMembership, channel: SlackChannel)`
or
`searchMessagesHoweverPossible(slackTeam: SlackTeam, slackChannel: SlackChannel)`

File: kifi-backend/modules/shoebox/app/com/keepit/slack/SlackCommander.scala (hunk 2 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/SlackCommander.scala:137]: #taste

Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/SlackCommander.scala:137]:
```
+   val kifiBotTokenOpt = for {
+      slackTeamId <- slackTeamIdOpt
+      slackTeam <- db.readOnlyMaster { implicit session =>
+        slackTeamRepo.getBySlackTeamId(slackTeamId)
+      }
+      kifiBotToken <- slackTeam.kifiBotToken
+    } yield kifiBotToken
+
+    kifiBotTokenOpt.fold(Future.successful(Set.empty)) { kifiBotToken =>
+      slackClient.validateToken(kifiBotToken).imap { isValid =>
+        if (isValid) SlackAuthScope.inheritableBotScopes else Set.empty
+      }
+    }
```

so pretty

**Pull request merged**